crates/ty_python_semantic/resources/mdtest/type_compendium/none.md

```diff
@@ -0,0 +1,77 @@
+# `None`
+
+## `None` as a singleton type
+
+The type `None` (or `NoneType`, see below) is a singleton type that has only one inhabitant: the
+object `None`.
+
+```py
+from ty_extensions import static_assert, is_singleton, is_equivalent_to
+
+n: None = None
+
+static_assert(is_singleton(None))
+```
+
+Just like for other singleton types, the only subtypes of `None` are `None` itself and `Never`:
+
+```py
+from ty_extensions import static_assert, is_subtype_of
+from typing_extensions import Never
+
+static_assert(is_subtype_of(None, None))
+static_assert(is_subtype_of(Never, None))
+```
+
+## Relationship to `Optional[T]`
+
+The type `Optional[T]` is an alias for `T | None` (or `Union[T, None]`):
+
+```py
+from ty_extensions import static_assert, is_equivalent_to
+from typing import Optional, Union
+
+class T: ...
+
+static_assert(is_equivalent_to(Optional[T], T | None))
+static_assert(is_equivalent_to(Optional[T], Union[T, None]))
+```
+
+## Type narrowing using `is`
+
+Just like for other singleton types, we support type narrowing using `is` or `is not` checks:
+
+```py
+from typing_extensions import assert_type
+
+class T: ...
+
+def f(x: T | None):
+    if x is None:
+        assert_type(x, None)
+    else:
+        assert_type(x, T)
+
+    assert_type(x, T | None)
+
+    if x is not None:
+        assert_type(x, T)
+    else:
+        assert_type(x, None)
+```
+
+## `NoneType`
+
+The form `NoneType` is only available since Python 3.10. `NoneType` is equivalent to `None`:
```

Comment:
We could say a bit more here about the fact that using `None` (which is an instance at runtime) to spell a type is a slightly weird special case, which only works because `NoneType` is a singleton type, so the name of the singleton instance can be used interchangeably with the name of the type. (For any other type, you'd use the name of the type, that is `NoneType`.) But I also don't think it's important that we add this.